src/agent/WMICommunication.cpp

```diff
@@ -0,0 +1,204 @@
+#include "WMICommunication.h"
+#include "WMICommunication.h"
+#include "WMICommunication.h"
+
+#define _WIN32_DCOM
+#include <iostream>
+#include <comdef.h>
+#pragma comment(lib, "wbemuuid.lib")
+
+
+WMICommunication::~WMICommunication()
+{
+    initialLocatorToWMI->Release();
+    services->Release();
+    pEnumerator->Release();
+    CoUninitialize();
+}
+
+bool WMICommunication::WMIInit()
+{
+    HRESULT hres = CoInitializeEx(0, COINIT_MULTITHREADED);
+
+    if (FAILED(hres))
+    {
+        throw std::exception("Failed to initialize COM library. Error code = 0x");
+        return false;
+    }
+
+    hres = CoInitializeSecurity(
+        NULL,
+        -1,                          // COM authentication
+        NULL,                        // Authentication services
+        NULL,                        // Reserved
+        RPC_C_AUTHN_LEVEL_PKT,       // Default authentication 
+        RPC_C_IMP_LEVEL_IMPERSONATE, // Default Impersonation  
+        NULL,                        // Authentication info
+        EOAC_NONE,                   // Additional capabilities 
+        NULL                         // Reserved
+    );
+
+
+    if (FAILED(hres))
+    {
+        CoUninitialize();
+        throw std::exception("Failed to initialize security. Error code = 0x");
+        return false;
+    }
+
+
+
+    hres = CoCreateInstance(
+        CLSID_WbemLocator,
+        0,
+        CLSCTX_INPROC_SERVER,
+        IID_IWbemLocator, (LPVOID*)&initialLocatorToWMI);
```

Comment:
initialLocatorToWMI is a shared pointer object. Casting pointer to it to void* and passing to CoCreateInstance seems like a serious mistake.

Reply:
We will talk about it.